feat(decode): Added support for passing options to the json decoder

svan-jansson opened the pull request:
There is currently no way of passing additional options to the json decoder (Poison). Such an option could be keys: :atoms.

This feature makes it possible to do this: `Neuron.Config.set(json_response_opts: [keys: :atoms])`

I've added a doctest for the Config setter and a response test confirming that the data returned is formatted according to the passed json_response_opts. All tests are passing.

Commit:
Added support for passing options to the json decoder

Commit:
Now unsetting the Neuron config after handling the result

uesteibar commented on lib/neuron/config.ex:
```diff
@@ -34,20 +34,26 @@ defmodule Neuron.Config do
   def set(context, headers: value), do: Store.set(context, :neuron_headers, value)
   def set(context, connection_opts: value), do: Store.set(context, :neuron_connection_opts, value)
 
+  def set(context, json_response_opts: value),
```
I'd rather call it `parse_options`, is that ok for you?

svan-jansson replied:
I agree that parse_options is better than json_response_opts. I'll make the change.

uesteibar reviewed (commented):
Just a tiny proposal, other than that this is very neat! Thanks for taking the time ❤️

Commit:
Renamed json_response_opts to parse_options

uesteibar commented:
Hey thanks for the quick reaction!

Just one more thing now that I gave it a thought, it would be great if this option could also be overwritten on [`Neuron.query/3`](https://github.com/uesteibar/neuron/blob/master/lib/neuron.ex#L92). Could you give this a try and see if we can make it work? I believe it's important to keep the ability to overwrite all options in `Neuron.Config` through the `options` parameter.

Commit:
Added parse_options as a function parameter and moved config getter

svan-jansson commented:
> 
> 
> Hey thanks for the quick reaction!
> 
> Just one more thing now that I gave it a thought, it would be great if this option could also be overwritten on [`Neuron.query/3`](https://github.com/uesteibar/neuron/blob/master/lib/neuron.ex#L92). Could you give this a try and see if we can make it work? I believe it's important to keep the ability to overwrite all options in `Neuron.Config` through the `options` parameter.

I now refactored the parse_options getter, to follow the convention already established in [`Neuron.query/3`](https://github.com/uesteibar/neuron/blob/master/lib/neuron.ex#L92). My change removed [`Response.handle/1`](https://github.com/uesteibar/neuron/blob/master/lib/neuron/response.ex#L15) and introduced `Response.handle/2`, that takes the resolved parse_options, that now can come from either the passed options or the `Config`. I updated all tests that referenced `Response.handle/1`.

This would be nice to have for the `connection_opts` too, which currently doesn't support this pattern. But that should probably be another PR.

uesteibar commented on lib/neuron.ex:
```diff
@@ -116,6 +116,11 @@ defmodule Neuron do
   end
 
+  defp handle_response(response, options) do
+    parse_options = parse_options(options)
```
Could we change this variable name to `parsed_options` to differentiate it from the function?

svan-jansson replied:
Sure thing. Fixed it!

uesteibar reviewed (commented):
Yay!! Great work here @svan-jansson 🎉 🎉 🎉 

One more tiny thing and this is good to go 😃

Commit:
Renamed variable to avoid naming confusion with function

uesteibar reviewed (approved):
Great job @svan-jansson!!

I'll merge this to master now and I'll released together with https://github.com/uesteibar/neuron/pull/27 once that one is done.

Thanks for taking the time to work on this! ❤️

svan-jansson commented:
Thank you for your quick feedback!

uesteibar commented:
This is now released as part of [v1.1.0](https://github.com/uesteibar/neuron/blob/master/CHANGELOG.md#110-2018-12-25).

Thank you and Marry Christmas! 🎅

svan-jansson commented:
Nice! Feliz Navidad!
